upload: Fix stacking of progress bar on canceling the upload.

shubhamdhama opened the pull request:
Here we can see on canceling first upload there is no stacking of progress bar while second upload. 
![gif fix upload](https://user-images.githubusercontent.com/22238472/38753328-9ca86e66-3f7b-11e8-89ad-153c9dc64717.gif)

I'll push second fix here soon.

shubhamdhama commented on static/js/upload.js:
```diff
@@ -167,7 +170,7 @@ exports.options = function (config) {
         raw_droppable: ['text/uri-list', 'text/plain'],
         drop: uploadStarted,
-        progressUpdated: progressUpdated,
+        globalProgressUpdated: globalProgressUpdated,
```
I've pushed another commit, in which we now use `globalPrgressUpdated` to track the overall progress of the all the files, which solves the dangling progress bar when many files are uploaded at a time. I've seen some minor delays in completing and inserting of link in composebox so it would be better to do some manual testing before merging this commit as I'm not sure whether it's my system or it happens before as well.

shubhamdhama commented:
Thanks, Joshua for reviewing

(I've also observed that the filedrop plugin we use and currently available at https://github.com/weixiyen/jquery-filedrop have many changes like we don't have `uploadStarted` function available to us so that we can track each upload(by each upload I mean I treat all the uploaded files as a single unit(`files`)). This can be helpful in solving the issue for separate progress bars when there are many different uploads at the same time).

timabbott commented:
Yeah, we should upgrade; the main challenge is that we'll need to carefully rebase or drop the patches we have to that library.

shubhamdhama commented:
* I've changed my logic and added some commits here, and it didn't require upgrading file-drop plugin as I simply made the upstream change to our library
![demo 1 compres](https://user-images.githubusercontent.com/22238472/38767488-f70a60ce-3fff-11e8-9771-a1e2edf41a22.gif)
* Now progress bar track each file individually(I've reverted the changes done in https://github.com/zulip/zulip/pull/9088#pullrequestreview-112152366)
* How to cause `uploadError` call, I haven't tested changes for it.
* I've used `file.lastModified` to track each file individually as there isn't any other valid id to uniquely identify each file. Is this okay?

shubhamdhama commented:
Rebased(waiting for tests getting passed). 
But I've found a bug unrelated to this PR(most probably) that when we have made many uploads at a time then only one upload get canceled (i.e. last one) and all other uploads keep continuing. I've reproduced this in master.
![peek 2018-04-16 23-23](https://user-images.githubusercontent.com/22238472/38826555-cafa07b2-41cd-11e8-8bcd-7f17f68e014f.gif)

shubhamdhama commented:
@timabbott Actually I didn't get your point totally as I'm not able to reproduce it, but my understanding here is that progress bar doesn't appear(virtually) on very fast upload so for that case I've set setTimeout for all hiding of progress bar. I've made this change in a separate commit. Is this understanding right, and does last commit makes sense now?

timabbott commented:
Can you get another developer to spend a bit of time playing with this and trying to reproduce the issues I saw?

shubhamdhama commented:
@HarshitOnGitHub can you try this PR if this works fine for you?

timabbott commented:
I can't reproduce issues anymore either, so it must have been some caching thing.  Merged, thanks @shubhamdhama!
